openssl/src/ssl/mod.rs

```diff
@@ -2140,6 +2223,48 @@ impl SslSessionRef {
         unsafe { SSL_SESSION_get_master_key(self.as_ptr(), buf.as_mut_ptr(), buf.len()) }
     }
 
+    /// Sets the master key
+    ///
+    /// This could be used to set up a session-based PSK.  You must ensure the length of the key is
+    /// suitable for the ciphersuite associated with the session.
+    #[corresponds(SSL_SESSION_set1_master_key)]
+    pub fn set_master_key(&mut self, key: &[u8]) -> Result<(), ErrorStack> {
```

Comment:
Sessions are reference counted so it's not sound to directly mutate them. We'll need to define a session builder like we have for SslContext instead.

Reply:
OK

Reply:
In 4565f0f36fd28b0323b49b488796c1ea48ea0259 I added `SslContextBuilder` modeled on the other builders.